File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/ExplorerAdapter.java
package app.editors.manager.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.explorer.UploadFile;
import app.editors.manager.mvp.models.list.Footer;
import app.editors.manager.mvp.models.list.Header;
import app.editors.manager.ui.adapters.base.BaseAdapter;
import app.editors.manager.ui.adapters.holders.BaseViewHolderExplorer;
import app.editors.manager.ui.adapters.holders.FooterViewHolder;
import app.editors.manager.ui.adapters.holders.UploadFileViewHolder;
import app.editors.manager.ui.adapters.holders.factory.TypeFactory;
import lib.toolkit.base.managers.utils.StringUtils;
import lib.toolkit.base.managers.utils.UiUtils;

public class ExplorerAdapter extends BaseAdapter<Entity> {

    @Inject
    public Context mContext;

    @Inject
    public PreferenceTool mPreferenceTool;

    private boolean mIsSelectMode;
    private boolean mIsFoldersMode;
    private boolean mIsFooter;
    private boolean mIsRoot;
    private boolean mIsSectionMy;

    private Footer mFooter;
    private TypeFactory mFactory;


    public ExplorerAdapter(TypeFactory factory) {
        App.getApp().getAppComponent().inject(this);
        mIsSelectMode = false;
        mIsFoldersMode = false;
        mIsFooter = false;
        mIsSectionMy = false;
        mFactory = factory;
        mFooter = new Footer();
    }

    @NonNull
    @Override
    public BaseViewHolderExplorer onCreateViewHolder(ViewGroup viewGroup, int typeHolder) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(typeHolder, viewGroup, false);
        return mFactory.createViewHolder(view, typeHolder, this);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof FooterViewHolder) {
            FooterViewHolder viewHolder = (FooterViewHolder) holder;
            viewHolder.bind(mFooter);
        } else {
            BaseViewHolderExplorer viewHolder = (BaseViewHolderExplorer) holder;
            setFileFavoriteStatus(position);
            viewHolder.bind(mList.get(position));
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position, List payloads) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position);
        } else {
            if (holder instanceof UploadFileViewHolder) {
                UploadFileViewHolder viewHolder = (UploadFileViewHolder) holder;
                viewHolder.updateProgress((UploadFile) payloads.get(0));
            }
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == getItemCount() - 1) {
            return FooterViewHolder.LAYOUT;
        } else {
            return getItemList().get(position).getType(mFactory);
        }
    }


    public void setSelectMode(final boolean isSelectMode) {
        mIsSelectMode = isSelectMode;
        notifyDataSetChanged();
    }

    public void setFoldersMode(final boolean isFoldersMode) {
        mIsFoldersMode = isFoldersMode;
        notifyDataSetChanged();
    }

    public void isLoading(final boolean isShow) {
        mIsFooter = isShow;
        notifyItemChanged(getItemCount() - 1);
    }

    public UploadFile getUploadFileById(String id) {
        if (mList != null) {
            for (Entity file : mList) {
                if (file instanceof UploadFile) {
                    UploadFile uploadFile = (UploadFile) file;
                    if (uploadFile.getId().equals(id)) {
                        return uploadFile;
                    }
                }
            }
        }
        return null;
    }

    public void removeUploadItemById(String id) {
        if (mList != null) {
            for (Entity file : mList) {
                if (file instanceof UploadFile) {
                    UploadFile uploadFile = (UploadFile) file;
                    if (uploadFile.getId().equals(id)) {
                        int position = mList.indexOf(uploadFile);
                        mList.remove(uploadFile);
                        notifyItemRemoved(position);
                        break;
                    }
                }
            }
        }
    }

    private void setFileFavoriteStatus(int position) {
        if(mList.get(position) instanceof File){
            File file = ((File)mList.get(position));
            if(!file.getFileStatus().isEmpty()) {
                int favoriteMask = Integer.parseInt(file.getFileStatus()) & Api.FileStatus.FAVORITE;
                if (favoriteMask != 0) {
                    file.setFavorite(true);
                } else {
                    file.setFavorite(false);
                }
            }
        }
    }

    public void setRoot(boolean root) {
        mIsRoot = root;
    }

    public void setSectionMy(boolean sectionMy) {
        mIsSectionMy = sectionMy;
    }

    public boolean isSelectMode() {
        return mIsSelectMode;
    }

    public boolean isFoldersMode() {
        return mIsFoldersMode;
    }

    public boolean isFooter() {
        return mIsFooter;
    }

    public boolean isRoot() {
        return mIsRoot;
    }

    public boolean isSectionMy() {
        return mIsSectionMy;
    }

    public void setFileIcon(final AppCompatImageView view, final String ext) {
        final StringUtils.Extension extension = StringUtils.getExtension(ext);
        @DrawableRes int resId = R.drawable.ic_type_file;
        @ColorRes int colorId = R.color.colorGrey;
        switch (extension) {
            case DOC:
                resId = R.drawable.ic_type_text_document;
                colorId = R.color.colorDocTint;
                break;
            case SHEET:
                resId = R.drawable.ic_type_spreadsheet;
                colorId = R.color.colorSheetTint;
                break;
            case PRESENTATION:
                resId = R.drawable.ic_type_presentation;
                colorId = R.color.colorPresentationTint;
                break;
            case IMAGE:
            case IMAGE_GIF:
                resId = R.drawable.ic_type_image;
                colorId = R.color.colorPicTint;
                break;
            case HTML:
            case EBOOK:
            case PDF:
                resId = R.drawable.ic_type_pdf;
                colorId = R.color.colorLightRed;
                break;
            case VIDEO_SUPPORT:
                resId = R.drawable.ic_type_video;
                colorId = R.color.colorBlack;
                break;
            case VIDEO:
                setAlphaIcon(view, R.drawable.ic_type_video);
                return;
            case ARCH:
                setAlphaIcon(view, R.drawable.ic_type_archive);
                return;
            case UNKNOWN:
                setAlphaIcon(view, R.drawable.ic_type_file);
                return;
        }

        view.setImageResource(resId);
        view.setAlpha(1.0f);
        view.setColorFilter(ContextCompat.getColor(mContext, colorId));
    }

    public void setAlphaIcon(final AppCompatImageView view, @DrawableRes final int resId) {
        view.setImageResource(resId);
        view.setAlpha(UiUtils.getFloatResource(mContext, R.dimen.alpha_medium));
        view.clearColorFilter();
    }

    public void setFolderIcon(final AppCompatImageView view, final Folder folder) {
        @DrawableRes int resId = R.drawable.ic_type_folder;
        if (folder.getShared() && folder.getProviderKey().isEmpty()) {
            resId = R.drawable.ic_type_folder_shared;
        } else if (isRoot() && folder.getProviderItem() && !folder.getProviderKey().isEmpty()) {
            switch (folder.getProviderKey()) {
                case Api.Storage.BOXNET:
                    resId = R.drawable.ic_storage_box;
                    break;
                case Api.Storage.DROPBOX:
                    resId = R.drawable.ic_storage_dropbox;
                    break;
                case Api.Storage.SHAREPOINT:
                    resId = R.drawable.ic_storage_sharepoint;
                    break;
                case Api.Storage.GOOGLEDRIVE:
                    resId = R.drawable.ic_storage_google;
                    break;
                case Api.Storage.ONEDRIVE:
                case Api.Storage.SKYDRIVE:
                    resId = R.drawable.ic_storage_onedrive;
                    break;
                case Api.Storage.YANDEX:
                    resId = R.drawable.ic_storage_yandex;
                    break;
                case Api.Storage.WEBDAV:
                    resId = R.drawable.ic_storage_webdav;
                    view.setImageResource(resId);
                    view.setAlpha(UiUtils.getFloatResource(mContext, R.dimen.alpha_medium));
                    return;
            }

            view.setImageResource(resId);
            view.setAlpha(1.0f);
            view.clearColorFilter();
            return;
        }

        view.setImageResource(resId);
        view.setAlpha(UiUtils.getFloatResource(mContext, R.dimen.alpha_medium));
    }

    public void checkHeaders() {
        if (mList != null) {
            for (int i = 0; i < mList.size(); i++) {
                if (mList.get(i) instanceof Header) {
                    Header header = (Header) mList.get(i);
                    int position = mList.indexOf(header);
                    if (position + 1 < mList.size() - 1) {
                        if (mList.get(i + 1) instanceof Header) {
                            mList.remove(header);
                            notifyItemRemoved(position);
                        }
                    } else if (mList.lastIndexOf(header) == mList.size() - 1) {
                        mList.remove(header);
                        notifyItemRemoved(position);
                    }
                }
            }
        }
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/ExplorerAdapter.kt
package app.editors.manager.ui.adapters

import android.content.Context
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.ItemTouchHelper.DOWN
import androidx.recyclerview.widget.ItemTouchHelper.UP
import androidx.recyclerview.widget.RecyclerView
import app.documents.core.network.manager.models.base.Entity
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.CloudFolder
import app.documents.core.network.manager.models.explorer.Item
import app.editors.manager.app.App.Companion.getApp
import app.editors.manager.app.accountOnline
import app.editors.manager.managers.tools.PreferenceTool
import app.editors.manager.mvp.models.list.Footer
import app.editors.manager.mvp.models.list.Templates
import app.editors.manager.mvp.presenters.main.PickerMode
import app.editors.manager.ui.adapters.base.BaseAdapter
import app.editors.manager.ui.adapters.diffutilscallback.EntityDiffUtilsCallback
import app.editors.manager.ui.adapters.holders.BaseViewHolderExplorer
import app.editors.manager.ui.adapters.holders.explorer.GridFileViewHolder
import app.editors.manager.ui.adapters.holders.explorer.GridFolderViewHolder
import app.editors.manager.ui.adapters.holders.explorer.GridFooterViewHolder
import app.editors.manager.ui.adapters.holders.explorer.GridRoomViewHolder
import app.editors.manager.ui.adapters.holders.explorer.GridTemplateViewHolder
import app.editors.manager.ui.adapters.holders.explorer.ListFileViewHolder
import app.editors.manager.ui.adapters.holders.explorer.ListFolderViewHolder
import app.editors.manager.ui.adapters.holders.explorer.ListFooterViewHolder
import app.editors.manager.ui.adapters.holders.explorer.ListRoomViewHolder
import app.editors.manager.ui.adapters.holders.explorer.ListTemplateViewHolder
import app.editors.manager.ui.adapters.holders.explorer.TemplatesFolderViewHolder
import app.editors.manager.ui.adapters.holders.factory.TypeFactoryExplorer
import lib.toolkit.base.ui.adapters.factory.inflate
import java.util.Collections
import javax.inject.Inject

interface ExplorerTouchAdapter {

    fun onMove(from: Int, to: Int): List<Item>
    fun onMoveFinished()
}

interface AdapterState {

    val accountId: String?
    val sortBy: String?

    var currentFolderId: String
    var sectionType: Int
    var isRoot: Boolean
    var isFooter: Boolean
    var isTrash: Boolean
    var isIndexing: Boolean
    var isGridView: Boolean
}

class ExplorerAdapter(
    private val factory: TypeFactoryExplorer,
    initialGridView: Boolean
) : BaseAdapter<Entity>(), AdapterState, ExplorerTouchAdapter {

    companion object {

        fun getTouchHelperCallback(
            adapter: ExplorerTouchAdapter,
            onMove: (items: List<Item>) -> Unit
        ): ItemTouchHelper.SimpleCallback {
            return object : ItemTouchHelper.SimpleCallback(UP or DOWN, 0) {

                override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    target: RecyclerView.ViewHolder
                ): Boolean {
                    val items = adapter.onMove(
                        viewHolder.absoluteAdapterPosition,
                        target.absoluteAdapterPosition
                    )
                    onMove(items)
                    return false
                }

                override fun onSwiped(
                    viewHolder: RecyclerView.ViewHolder,
                    direction: Int
                ) {
                }

                override fun clearView(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder
                ) {
                    super.clearView(recyclerView, viewHolder)
                    adapter.onMoveFinished()
                }
            }
        }
    }

    @Inject
    lateinit var context: Context

    @Inject
    lateinit var preferenceTool: PreferenceTool

    override val accountId: String? by lazy { context.accountOnline?.id }

    override var currentFolderId: String = ""
    override var sectionType: Int = -1
    override var isRoot: Boolean = false
    override var isFooter: Boolean = false
    override var isTrash: Boolean = false
    override var isIndexing: Boolean = false
    override var isGridView: Boolean = initialGridView

    override val sortBy: String? get() = preferenceTool.sortBy

    var onDragStartListener: (RecyclerView.ViewHolder) -> Unit = {}

    var isSelectMode = false
        set(isSelectMode) {
            field = isSelectMode
            notifyItemRangeChanged(0, itemCount - 1, ExplorerPayload.SELECTION)
        }

    var pickerMode: PickerMode = PickerMode.None

    private val footer: Footer = Footer()

    init {
        getApp().appComponent.inject(this)
    }

    fun updateItems(list: List<Entity>?, currentId: String? = null) {
        if (list == null) return
        val listCurrentId = currentId ?: currentFolderId
        if (itemList != null && listCurrentId == currentFolderId) {
            val callback = EntityDiffUtilsCallback(list, itemList)
            val result = DiffUtil.calculateDiff(callback)
            super.setData(list)
            result.dispatchUpdatesTo(this)
        } else {
            currentFolderId = listCurrentId
            super.setItems(list)
        }
    }

    fun updateItemById(id: String, payload: ExplorerPayload) {
        val position = itemList.indexOfFirst { it.getEntityId() == id }
        if (position != -1) notifyItemChanged(position, payload)
    }

    override fun onCreateViewHolder(view: ViewGroup, type: Int): BaseViewHolderExplorer<*> {
        return factory.createViewHolder(view.inflate(type), type, this)
    }

    @Suppress("UNCHECKED_CAST")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ListFooterViewHolder) {
            holder.bind(footer)
        } else {
            (holder as BaseViewHolderExplorer<Entity>).bind(mList[position])
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int,
        payloads: List<*>
    ) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
            return
        }
        for (payload in payloads) {
            when (payload) {
                ExplorerPayload.SELECTION -> {
                    (holder as? BaseViewHolderExplorer<Entity>)?.bindSelection(mList[position])
                }
                ExplorerPayload.THUMBNAIL -> {
                    if (!isGridView) return
                    (mList[position] as? CloudFile?)?.let { file ->
                        (holder as? GridFileViewHolder)?.bindThumbnail(file)
                    }
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + 1
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == itemCount - 1) {
            if (isGridView) GridFooterViewHolder.LAYOUT else ListFooterViewHolder.LAYOUT
        } else {
            when (val item = itemList[position]) {
                is CloudFile -> if (isGridView) GridFileViewHolder.LAYOUT else ListFileViewHolder.LAYOUT
                is CloudFolder -> getFolderLayout(item)
                Templates -> TemplatesFolderViewHolder.LAYOUT
                else -> 0
            }
        }
    }

    private fun getFolderLayout(item: CloudFolder): Int {
        return if (isGridView) {
            when {
                item.isTemplate -> GridTemplateViewHolder.LAYOUT
                item.isRoom -> GridRoomViewHolder.LAYOUT
                else -> GridFolderViewHolder.LAYOUT
            }
        } else {
            when {
                item.isTemplate -> ListTemplateViewHolder.LAYOUT
                item.isRoom -> ListRoomViewHolder.LAYOUT
                else -> ListFolderViewHolder.LAYOUT
            }
        }
    }

    fun isLoading(isShow: Boolean) {
        isFooter = isShow
        notifyItemChanged(itemCount - 1)
    }

    override fun onMove(from: Int, to: Int): List<Item> {
        val lastIndex = itemList.lastIndex
        if (from !in 0..lastIndex || to !in 0..lastIndex) return emptyList()
        Collections.swap(itemList, from, to)
        notifyItemMoved(from, to)
        swapIndexes(from, to)
        return listOf(itemList[from] as Item, itemList[to] as Item)
    }

    override fun onMoveFinished() {
        notifyDataSetChanged()
    }

    private fun swapIndexes(from: Int, to: Int) {
        val list = itemList
        val indexFrom = (list[from] as Item).index
        val indexTo = (list[to] as Item).index
        (list[from] as Item).index = indexTo
        (list[to] as Item).index = indexFrom
    }

    enum class ExplorerPayload {
        SELECTION, THUMBNAIL
    }
}